### ISSUE TITLE
array_hash fails on reversed arrays

### ISSUE THREAD

**Issue by arvoelke** [nengo/nengo#1122]
Alternative bug title: "Cannot use `PresentInput` process with a reversed array".

```
do_bug = True      # set to True to use array that will bug on array_hash
fix_bug = False    # set to True to fix the array so that it can be hashed
check_bug = False  # set to True to invoke the bugged code directly

if do_bug:
    inputs = np.asarray([0, 1, 2])[::-1]
else:
    inputs = np.asarray([0, 1, 2])

if fix_bug:
    inputs = np.array(inputs)

if check_bug:
    nengo.utils.numpy.array_hash(inputs)

with nengo.Network() as model:
    nengo.Node(output=PresentInput(inputs, 1.0))

with nengo.Simulator(model) as sim:
    sim.run(1.0)
```

```
c:\users\aaron\ctn\nengo\nengo\utils\numpy.pyc in array_hash(a, n)
     72         v = a.view()
     73         v.setflags(write=False)
---> 74         return hash(v.data if PY2 else v.data.tobytes())
     75     else:
     76         # pick random elements to hash

AttributeError: cannot get single-segment buffer for discontiguous array
```

Submitted on behalf of @celiasmith.

**Comment by hunse** [comment]
One workaround is to copy the reversed array: `inputs = np.array(np.array([0, 1, 2])[::-1])`

**Comment by arvoelke** [comment]
Yeah sorry for not annotating my code in text. That is what `fix_bug=True` does.

**Comment by tbekolay** [comment]
I suspect that `PresentInput` will be used with relative large arrays, so I'm not sure that we want to copy the array by default... though we could try/except and copy if it fails?

Also, FYI the error here has nothing to do with hashing, it's the `v.data` part. `v.data` is a Python [buffer](https://docs.python.org/2/c-api/buffer.html) that is the underlying data in the numpy array. If you try to do `v.data` on a view, it will raise this error if the array is not contiguous... You can check if an array is continuous with `v.flags`:

```ipython
In [1]: import numpy as np

In [2]: a = np.arange(10)

In [3]: a.flags
Out[3]: 
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  UPDATEIFCOPY : False

In [4]: a[::-1].flags
Out[4]: 
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  UPDATEIFCOPY : False

In [5]: a.data
Out[5]: <read-write buffer for 0x7f074ec658a0, size 80, offset 0 at 0x7f074ec2e330>

In [6]: a[::-1].data
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-e75ef4c0a759> in <module>()
----> 1 a[::-1].data

AttributeError: cannot get single-segment buffer for discontiguous array
```

**Comment by hunse** [comment]
Does this actually fail for large arrays? `array_hash` uses a random selection method for large arrays, it doesn't hash the whole thing. So this might actually be specific to smaller arrays.

**Comment by jgosmann** [comment]
I can't reproduce this with Python 3.5.3, Nengo 2.4.0, and Numpy 1.13.0.

**Comment by arvoelke** [comment]
This is still happening to me in TravisCI when testing against `python=2.7`, the current master branch of Nengo, and in much more general situations as well (e.g., setting the `transform` on a connection). Here is a minimal reproducer:

```
cd nengo
conda create --name=py27 python=2.7 numpy=1.10 scipy=0.17.0
conda activate py27
python setup.py -q install
```

```python
import numpy as np
import nengo

with nengo.Network() as model:
    nengo.Connection(
        nengo.Node(size_in=1),
        nengo.Node(size_in=1),
        transform=np.zeros(1)[:, None])

with nengo.Simulator(model) as sim:
    pass
```

```
Traceback (most recent call last):
  File "temp.py", line 10, in <module>
    with nengo.Simulator(model) as sim:
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/simulator.py", line 160, in __init__
    progress=pt.next_stage("Building", "Build"))
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/builder/builder.py", line 123, in build
    built = self.builder.build(self, obj, *args, **kwargs)
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/builder/builder.py", line 218, in build
    return cls.builders[obj_cls](model, obj, *args, **kwargs)
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/builder/network.py", line 113, in build_network
    model.build(conn)
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/builder/builder.py", line 123, in build
    built = self.builder.build(self, obj, *args, **kwargs)
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/builder/builder.py", line 218, in build
    return cls.builders[obj_cls](model, obj, *args, **kwargs)
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/builder/connection.py", line 261, in build_connection
    rng=rng)
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/builder/builder.py", line 123, in build
    built = self.builder.build(self, obj, *args, **kwargs)
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/builder/builder.py", line 206, in build
    if model.has_built(obj):
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/builder/builder.py", line 144, in has_built
    return obj in self.params
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/params.py", line 527, in __hash__
    p.hashvalue(self) for p in self._params)))
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/params.py", line 527, in <genexpr>
    p.hashvalue(self) for p in self._params)))
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/params.py", line 392, in hashvalue
    return array_hash(self.__get__(instance, None))
  File "/home/arvoelke/anaconda3/envs/py27/lib/python2.7/site-packages/nengo-2.8.1.dev0-py2.7.egg/nengo/utils/numpy.py", line 74, in array_hash
    return hash(v.data if PY2 else v.data.tobytes())
AttributeError: cannot get single-segment buffer for discontiguous array
```

Upgrading `numpy` to `1.13.0` (as Jan observed above) fixes the problem.

**Comment by hunse** [comment]
As I said above, I think this will only happen in the small-array case. One option would be to copy arrays in that case only.

The other option, of course, is to force newer Numpy.

**Comment by arvoelke** [comment]
Good point -- I've pushed the copy fix in #1494.

**Comment by arvoelke** [comment]
Closing as we no longer support Python 2.7.
